### PR TITLE
create test for internal/errors/errors.go

### PR DESCRIPTION
**Pull request by vankichi**
Signed-off-by: vankichi <kyukawa315@gmail.com>

I created the unit test for `internal/errors/errors.go`.
And refactored the functions of `internal/errors/errors.go` which are `ErrOptionFailed`, `ErrLoggingRetry`, and `ErrLoggingFailed`.

also, grammar check by grammarly is completed.

### Environment:

- Go Version: 1.15.2
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [x] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/errors/errors_test.go [hunk 1]
```diff
@@ -0,0 +1,1407 @@
...
+		},
+		{
+			name: "return an ErrOptionFailed error when err is not empty and ref is empty.",
+			args: args{
+				err: errors.New("option failed error"),
+			},
+			want: want{
+				errors.New("failed to setup option :\t: option failed error"),
+			},
+		},
```
Comment (hlts2) [inline internal/errors/errors_test.go:191]: Should I add pattern that `runtime.FuncForPC(ref.Pointer()).Name())`  is not empty ?
if I made a mistake, I am sorry.

File: internal/errors/errors_test.go [hunk 2]
```diff
@@ -0,0 +1,1407 @@
...
+		if !Is(got, w.want) {
+			return fmt.Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
+		}
+		return nil
+	}
+	tests := []test{
+		{
+			name: "return an unwrapped error when err is not empty.",
+			args: args{
+				err: errors.New("err"),
```
Comment (hlts2) [inline internal/errors/errors_test.go:943]: Should I use wrapped error insted of normal error?

File: internal/errors/errors_test.go (hunk 1 above)
Comment (vankichi) [reply internal/errors/errors_test.go:191]: well, it would be better.
how should I create it?

File: internal/errors/errors_test.go [hunk 3]
```diff
@@ -0,0 +1,1499 @@
...
+	tests := []test{
+		{
+			name: "return true when err and target is not empty.",
+			args: args{
+				err:    errors.New("err"),
+				target: errors.New("err is occurred"),
+			},
+			want: want{
+				true,
+			},
```
[the remark this answers is not in the document]
Comment (vankichi) [reply internal/errors/errors_test.go:1477]: @hlts2 
Your comment will be fine at point of another view.
I think this function `As` is a just alias of `As()` function. which is golang original errors package,  in other word, we do not implement this function.
IMO, we should create many patterns tests.
What do you think about it?
cc: @kevindiu

[the remark this answers is not in the document]
Comment (hlts2) [reply internal/errors/errors_test.go:1476]: Thank you for your comment.
And I got it. I agree with you 👍 
so my comment was mistake. :bow: I think this code is ok!
I am sorry.

**Review by hlts2 — APPROVED** [review]
LGTM

File: internal/errors/errors_test.go [hunk 4]
```diff
@@ -0,0 +1,1498 @@
...
+			return test{
+				name: "return an ErrOptionFailed error when err is not empty and ref is empty.",
+				args: args{
+					err: errors.New("option failed error"),
+				},
+				want: want{
+					wantErr,
+				},
+			}
+		}(),
```
Comment (kevindiu) [inline internal/errors/errors_test.go:213]: can you add a test case that both `ref` and `err` are empty?

File: internal/errors/errors_test.go [hunk 5]
```diff
@@ -0,0 +1,1498 @@
...
+					ref: func() reflect.Value {
+						var i int
+						return reflect.ValueOf(&i)
+					}(),
+				},
+				want: want{
+					wantErr,
+				},
+			}
+		}(),
```
Comment (kevindiu) [inline internal/errors/errors_test.go:534]: can you add a test case that both ref and err are empty?

File: internal/errors/errors_test.go [hunk 6]
```diff
@@ -0,0 +1,1498 @@
...
```
Comment (kevindiu) [inline internal/errors/errors_test.go:622]: can you add a test case that both ref and err are empty?

File: internal/errors/errors.go [hunk 7]
```diff
@@ -26,48 +26,71 @@ import (
...
+	// When the message is nil, it will return nil instead of an error.
 	New = func(msg string) error {
 		if msg == "" {
 			return nil
 		}
 		return errors.New(msg)
 	}
 
+	// Wrap represents a function to generate an error that is used by input error and message.
+	// When both of the input is nil, it will return a new error with the message even message is nil.
```
Comment (kevindiu) [inline internal/errors/errors.go:92]: I think it is incorrect, the nil will be returned if the message is nil (line 85)
```suggestion
	// When both of the input are nil, it will return an error when the error message is not empty.
```

File: internal/errors/errors_test.go [hunk 8]
```diff
@@ -0,0 +1,1498 @@
...
+	}
+	defaultCheckFunc := func(w want, got error) error {
+		if !Is(got, w.want) {
+			return fmt.Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
+		}
+		return nil
+	}
+	tests := []test{
+		func() test {
+			wantErr := errors.Unwrap(Wrap(errors.New("err"), "invalid parameter"))
```
Comment (kevindiu) [inline internal/errors/errors_test.go:1020]: Sorry I am not understand what is the actual `wantErr` is. Maybe this one is correct?
```suggestion
			wantErr := errors.New("err")
```

File: internal/errors/errors_test.go [hunk 9]
```diff
@@ -0,0 +1,1498 @@
...
+	defaultCheckFunc := func(w want, got error) error {
+		if !Is(got, w.want) {
+			return fmt.Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
+		}
+		return nil
+	}
+	tests := []test{
+		func() test {
+			wantErr := errors.New("err")
+			err := fmt.Errorf("%s: %w", "error occurs", wantErr)
```
Comment (kevindiu) [inline internal/errors/errors_test.go:1084]: can you tell me why you use `pkg/errors` and `fmt` at the same time here?

File: internal/errors/errors_test.go (hunk 8 above)
Comment (vankichi) [reply internal/errors/errors_test.go:1020]: @kevindiu FYI: https://github.com/vdaas/vald/pull/929#discussion_r563425046

File: internal/errors/errors_test.go (hunk 4 above)
Comment (vankichi) [reply internal/errors/errors_test.go:213]: I already added

File: internal/errors/errors_test.go (hunk 5 above)
Comment (vankichi) [reply internal/errors/errors_test.go:534]: I already added

File: internal/errors/errors_test.go (hunk 6 above)
Comment (vankichi) [reply internal/errors/errors_test.go:622]: I already added

File: internal/errors/errors_test.go (hunk 9 above)
Comment (vankichi) [reply internal/errors/errors_test.go:1084]: @kevindiu here https://github.com/vdaas/vald/pull/929#discussion_r563425046

**Review by kevindiu — APPROVED** [review]
LGTM

File: internal/errors/errors_test.go [hunk 10]
```diff
@@ -0,0 +1,1499 @@
...
+		name       string
+		args       args
+		want       want
+		checkFunc  func(want, error) error
+		beforeFunc func(args)
+		afterFunc  func(args)
+	}
+	defaultCheckFunc := func(w want, got error) error {
+		if !Is(got, w.want) {
+			fmt.Println(got)
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply internal/errors/errors_test.go:165]:
```suggestion
```

**Review by kevindiu — APPROVED** [review]
LGTM
